### PR TITLE
adds composer require command

### LINKED ISSUE #2: Add composer require command
Description can be taken from the issue here https://github.com/prooph/micro/issues/21, except this is `composer require` and not `composer install/update`.

### PR DESCRIPTION
**Pull request by oqq**
This PR closes #2 

Since the require command needs a package as argument, I have had to copy the abstract composer command content. Other implementations would patch around the current abstract implementation and was to messy I guess. Also with unit tests.

Another way would be to group `update` and `require` together, since booth could have a package as argument, but `install` has not.

But for now I am fine with that.

### CODE REVIEW

File: src/Command/ComposerRequireCommand.php [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+            $phpServices[$service] = [
+                'php_version' => $phpVersionMatches[1],
+            ];
+        }
+
+        return $phpServices;
+    }
+
+    private function getRequestedServices(InputInterface $input, OutputStyle $io, array $requestedServices): array
```
Comment (prolic) [inline src/Command/ComposerRequireCommand.php]: Last option should be declared services

File: tests/Command/ComposerRequireCommandTest.php [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+
+use PHPUnit\Framework\TestCase;
+use Prooph\MicroCli\Command\ComposerRequireCommand;
+use Symfony\Component\Console\Application;
+use Symfony\Component\Console\Tester\CommandTester;
+
+/**
+ * @coversDefaultClass \Prooph\MicroCli\Command\ComposerRequireCommand
+ */
+final class ComposerRequireCommandTest extends TestCase
```
Comment (prolic) [inline tests/Command/ComposerRequireCommandTest.php]: No interface, so not final

File: src/Command/ComposerRequireCommand.php [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+        $processBuilder = new ProcessBuilder();
+        $processBuilder->setPrefix($dockerExecutable);
+        $processBuilder->setTimeout($timeout);
+
+        /** @var ProcessHelper $processHelper */
+        $processHelper = $this->getHelper('process');
+
+        $package = $input->getArgument('package');
+
+        foreach ($requestedServices as $service => $values) {
```
Comment (prolic) [inline src/Command/ComposerRequireCommand.php:87]: This contains only one entry, right? So why the loop?

Comment (oqq) [reply src/Command/ComposerRequireCommand.php:87]: There is an "all" option like the other commands.

File: tests/Command/ComposerRequireCommandTest.php (hunk 2 above)
Comment (oqq) [reply tests/Command/ComposerRequireCommandTest.php]: Is that an code style rule for prooph? Maybe I have to change that on more parts.

I declare classes also final if they are DTOs or there is no need to mock they. Like test classes.

File: src/Command/ComposerRequireCommand.php (hunk 1 above)
Comment (oqq) [reply src/Command/ComposerRequireCommand.php]: done

**Comment by prolic** [comment]
@oqq http://ocramius.github.io/blog/when-to-declare-classes-final/

**Review by prolic — APPROVED** [review]

**Comment by oqq** [comment]
@prolic
> `final` should be used whenever possible

**Comment by prolic** [comment]
*TL;DR: Make your classes always final, if they implement an interface, and
no other public methods are defined*

Am 06.03.2017 18:13 schrieb "Eric Braun" <notifications@github.com>:

**Comment by oqq** [comment]
I guess thats the minimum when to declare classes final.
That should not be a `final` approach. ;)

I would prefer to declare classes always final where possible. And this is always true for not abstract unit test classes, DTOs and of course for classes which implements an interface.

**Comment by prolic** [comment]
Noone will extend a unit test, promised ;-)
Also all unit tests in prooph are not final, so let's have consistency here.

I agree with some other stuff being final, too, like enums, aggregate roots, value objects and so on.

**Comment by oqq** [comment]
I am fine with that if it is a prooph rule to be consistent and will fix it if I see some code in the wild. 👍

**Comment by prolic** [comment]
Thank you
